src/plugins/intel_cpu/src/transformations/tpp/x64/pass/fuse_tpp_to_equations.cpp

```diff
@@ -0,0 +1,119 @@
+        bool not_supported_op =
+                // ticket: 152532
+                ov::is_type<ov::snippets::op::ReduceBase>(n) ||
+                // ticket: 152510
+                ov::is_type<ov::op::v0::Relu>(n);
+        return not_supported_op ? nullptr : tpp;
+    };
+
+    // Note: we don't support exprs with more than 1 output yet. It's a technical limitation, but there are no use cases
+    const auto tpp_root = get_tpp_op(root);
+    if (!tpp_root || !supported_num_out(root->output(0)))
+        return false;
+
+    const auto root_subtensor = PortDescriptorUtils::get_port_descriptor_ptr(root->output(0))->get_subtensor();
+    auto supported_subtensor = [&root_subtensor](const snippets::VectorDims& subtensor) {
+        const auto size = subtensor.size();
+        if (size != root_subtensor.size())
+            return false;
+        for (size_t i = 0; i < size; i++) {
+            if (subtensor[i] != root_subtensor[i] && subtensor[i] != 1)
+                return false;
+        }
+        return true;
+    };
+
+    std::vector<op::OpDescTPP> op_types;
+    auto visit_node = [&](const Output<ov::Node>& out) {
+        const auto& node = out.get_node_shared_ptr();
+        const auto& pd = PortDescriptorUtils::get_port_descriptor_ptr(out);
+        auto tpp_expr = get_tpp_op(node);
+        if (!tpp_expr || !supported_num_out(out) || !supported_subtensor(pd->get_subtensor())) {
+            // Every skipped node is added to the descriptors as an argument
+            op_descs.emplace_back(op::OpDescTPP::ARITY::ZERO, eq_ivals.size());
+            eq_ivals.push_back(out);
+            return false;
+        }
+        op_descs.emplace_back(tpp_expr->get_op_desc());
+        node_replace_map.insert({node, nullptr});
+        return true;
+    };
+    OutputVector to_visit{root->output(0)};
+    while (!to_visit.empty()) {
+        const auto& current = to_visit.back();
+        to_visit.pop_back();
+        if (visit_node(current)) {
+            const auto& node_ivals = current.get_node_shared_ptr()->input_values();
+            to_visit.insert(to_visit.end(), node_ivals.rbegin(), node_ivals.rend());
+        }
+    }
+
+
+    auto equation = std::make_shared<op::EquationTPP>(eq_ivals, op_descs);
+
+    for (auto& kv : node_replace_map)
+        kv.second = equation;
+    replace_nodes(m, {}, node_replace_map);
+    for (const auto& in : equation->inputs()) {
+        ov::snippets::lowered::PortDescriptorUtils::set_port_descriptor(in, root_subtensor);
+    }
+    ov::snippets::lowered::PortDescriptorUtils::set_port_descriptor(equation->output(0), root_subtensor);
+    return true;
+}
+
+bool FuseTPPToEquations::run_on_model(const std::shared_ptr<ov::Model>& m) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::FuseTPPToEquations")
+    bool modified = false;
+    const auto& results = m->get_results();
+
+    NodeVector to_visit(results.begin(), results.end());
```

Comment:
Can we use `std::stack` here? This code is based on the classic `DFS` algorithm and uses `pop_back`, `back` and `push_back` methods - the ideal situation for the `std::stack` 😄 
It's valid for the same algorithm in `fuse_from_root`

Reply:
Moreover, we already have the helper `visit_path` for expressions and with `BFS` algorithm. What's about to add the similar helper with `DFS` for nodes? What do you think? Can it be useful?

Reply:
> This code is based on the classic DFS algorithm
I know, right? 😁 I would even say that it actually is a classic DFS, just like in textbooks

Regarding the `std::stack` - it's just a container adapter, it merely limits the underlying container's interface. 
But we don't really want to limit ourselves here, so I deliberately used std::vector for two reasons:
1) It's much easier to debug the process, since you can print to_visit easily
2) std::vector is more generic, since it'll allow the visitor to peek into the LIFO queue if necessary. We don't currently use it for eltwises, but it might be handy if we support Reduction ops for example (i.e. Reduction can be fused only if the second brunch, which was not visited yet, doesn't have broadcasting)

Reply:
Okey, thank you for the sharing of your vision!
Anyway I think that this code won't be updated for a long time and we won't need to debug it 😄 
However, I agree that `std::stack` doesn't provide new benefits compared to the current solution in general

Reply:
Discussed offline: decided to keep the current solution as more future-proofed.